server/app/auth/oidc/applicant/OidcApplicantProfileAdapter.java

```diff
@@ -0,0 +1,127 @@
+      Provider<UserRepository> applicantRepositoryProvider,
+      String emailAttributeName,
+      @Nullable String localeAttributeName,
+      ImmutableList<String> nameAttributeNames) {
+    super(configuration, client, profileFactory, applicantRepositoryProvider);
+    this.emailAttributeName = Preconditions.checkNotNull(emailAttributeName);
+    this.localeAttributeName = Optional.ofNullable(localeAttributeName);
+    this.nameAttributeNames = Preconditions.checkNotNull(nameAttributeNames);
+  }
+
+  protected Optional<String> getName(OidcProfile oidcProfile) {
+    String name =
+        nameAttributeNames.stream()
+            .filter(s -> !s.isBlank())
+            .map((String attrName) -> oidcProfile.getAttribute(attrName, String.class))
+            .filter(s -> !Strings.isNullOrEmpty(s))
+            .collect(Collectors.joining(" "));
+
+    return Optional.ofNullable(Strings.emptyToNull(name));
+  }
+
+  protected Optional<String> getLocale(OidcProfile oidcProfile) {
+    return localeAttributeName
+        .map(name -> oidcProfile.getAttribute(name, String.class))
+        .filter(s -> !Strings.isNullOrEmpty(s));
+  }
+
+  // Legacy. TODO: remove.
+  @Override
+  protected String emailAttributeName() {
+    return emailAttributeName;
+  }
+
+  /** Create a totally new Applicant CiviForm profile informed by the provided OidcProfile. */
+  @Override
+  public CiviFormProfile createEmptyCiviFormProfile(OidcProfile profile) {
+    return profileFactory.wrapProfileData(profileFactory.createNewApplicant());
+  }
+
+  protected boolean isTrustedIntermediary(CiviFormProfile profile) {
+    return profile.getAccount().join().getMemberOfGroup().isPresent();
+  }
+
+  @Override
+  protected ImmutableSet<Roles> roles(CiviFormProfile profile, OidcProfile oidcProfile) {
+    if (isTrustedIntermediary(profile)) {
+      return ImmutableSet.of(Roles.ROLE_APPLICANT, Roles.ROLE_TI);
+    }
+    return ImmutableSet.of(Roles.ROLE_APPLICANT);
+  }
+
+  @Override
+  protected void adaptForRole(CiviFormProfile profile, ImmutableSet<Roles> roles) {
+    // not needed
+  }
+
+  @Override
+  protected void possiblyModifyConfigBasedOnCred(Credentials cred) {
+    // not needed
+  }
+
+  /** Merge the two provided profiles into a new CiviFormProfileData. */
+  @Override
+  protected CiviFormProfileData mergeCiviFormProfile(
+      CiviFormProfile civiformProfile, OidcProfile oidcProfile) {
+    final String locale = getLocale(oidcProfile).orElse("");
+    final String name = getName(oidcProfile).orElse("");
+
+    if (!locale.isBlank() || !name.isBlank()) {
+      civiformProfile
+          .getApplicant()
+          .thenApplyAsync(
+              applicant -> {
+                if (!locale.isBlank()) {
+                  applicant.getApplicantData().setPreferredLocale(Locale.forLanguageTag(locale));
+                }
+                if (!name.isBlank()) {
+                  applicant.getApplicantData().setUserName(name);
+                }
```

Comment:
suggestion (blocking): this could be simplified by avoiding the empty strings

```suggestion
    Optional<String> maybeLocale = getLocale(oidcProfile);
    Optional<String> maybeName = getName(oidcProfile);

    if (maybeLocale.isPresent() || maybeName.isPresent()) {
      civiformProfile
          .getApplicant()
          .thenApplyAsync(
              applicant -> {
                maybeLocale.ifPresent(locale -> applicant.getApplicantData().setPreferredLocale(Locale.forLanguageTag(locale)));
                maybeName.ifPresent(name -> applicant.getApplicantData().setUserName(name));
```